backend/src/apiserver/server/util.go

```diff
@@ -216,64 +216,88 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
 	return nil
 }
 
-func ValidatePipelineSpec(resourceManager *resource.ResourceManager, spec *api.PipelineSpec) error {
-	if spec == nil || (spec.GetPipelineId() == "" && spec.GetWorkflowManifest() == "") {
-		return util.NewInvalidInputError("Please specify a pipeline by providing a pipeline ID or workflow manifest.")
+func ValidatePipelineSpecAndResourceReferences(resourceManager *resource.ResourceManager, spec *api.PipelineSpec, resourceReferences []*api.ResourceReference) error {
+	// centralize all the nil checks here
+	pipelineId := spec.GetPipelineId()
```

Comment:
Is the reason directly using `spec.GetPipelineId()` already checks if `spec` is nil?
Good to know, I just learned this is possible today.

Reply:
Yep. spec.GetPipelineId() already checks if spec is nil or not.